scripts/cnode-helper-scripts/cncli.sh

```diff
@@ -797,10 +807,203 @@ cncliPTsendslots() {
+
+  echo "Processing previous epoch: ${one_or_many}"
+  stake_param_prev="--active-stake ${active_stake_hist} --pool-stake ${pool_stake_hist}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --epoch="${one_or_many}" --ledger-set prev ${stake_param_prev} |
+  jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+  sed 's/"//g' >> "$tmpcsv"
+
+  return 0
+}
+
+processAllEpochs() {
+  getCurrNextEpoch
+  IFS=' ' read -r -a epochs_array <<< "$EPOCHS"
+
+  for EPOCH in "${epochs_array[@]}"; do
+    if ! getConsensus "${EPOCH}"; then echo "ERROR: Failed to fetch protocol parameters for epoch ${EPOCH}.; return 1"; fi
+     if [[ "$EPOCH" == "$curr_epoch" ]]; then
+       runCurrentEpoch
+     elif [[ "$EPOCH" == "$next_epoch" ]]; then
+       runNextEpoch
+     else
+       runPreviousEpochs
+     fi
+  done
+
+  id=1
+  while IFS= read -r row; do
+    echo "$id,$row" >> "$csvfile"
+    ((id++))
+  done < "$tmpcsv"
+
+  sqlite3 "$BLOCKLOG_DB" <<EOF
+DELETE FROM epochdata;
+VACUUM;
+.mode csv
+.import '$csvfile' epochdata
+REINDEX epochdata;
+EOF
+
+  row_count=$(sqlite3 "$BLOCKLOG_DB" "SELECT COUNT(*) FROM epochdata;")
+  echo "$row_count rows have been loaded into epochdata table in blocklog db"
+  echo "~ CNCLI epochdata table load completed ~"
+
+  rm $csvfile $tmpcsv
+}
+
+processSingleEpoch() {
+ if isNumber "$1"; then
+     getCurrNextEpoch
+     IFS=' ' read -r -a epochs_array <<< "$EPOCHS"
+
+   if [[ ! " ${epochs_array[@]} " =~ "$1" ]]; then
+     echo "No slots found in blocklog table for epoch ${EPOCH}."
+     echo
+     exit 1
+   fi
+
+   if ! getConsensus "${1}"; then echo "ERROR: Failed to fetch protocol parameters for epoch ${1}.; return 1"; fi
+
+   if [[ "$1" == "$curr_epoch" ]]; then
+      runCurrentEpoch
+   elif [[ "$1" == "$next_epoch" ]]; then
+      runNextEpoch
+   else
+      runPreviousEpochs
+   fi
+
+   ID=$(sqlite3 "$BLOCKLOG_DB" "SELECT max(id) + 1 FROM epochdata;")
+   csv_row=$(cat "$tmpcsv")
+   modified_csv_row="${ID},${csv_row}"
+   echo "$modified_csv_row" > "$onerow_csv"
+
+   sqlite3 "$BLOCKLOG_DB" "DELETE FROM epochdata WHERE epoch = '$1';"
+   sqlite3 "$BLOCKLOG_DB" <<EOF
+.mode csv
+.import "$onerow_csv" epochdata
+REINDEX epochdata;
+EOF
+   row_count=$(sqlite3 "$BLOCKLOG_DB" "SELECT COUNT(*) FROM epochdata WHERE epoch = '$1';")
+   echo "$row_count row has been loaded into epochdata table in blocklog db for epoch '$1'"
+   echo "~ CNCLI epochdata table load completed ~"
+   echo
+
+   rm $onerow_csv $tmpcsv
+ else
+   echo "No data found in $tmpcsv."
+   exit 1
+ fi
+}
+
+cncliEpochData() {
+  cncliInit
+  cncliParams="--db ${CNCLI_DB} --byron-genesis ${BYRON_GENESIS_JSON} --shelley-genesis ${GENESIS_JSON} --pool-id ${POOL_ID} --pool-vrf-skey ${POOL_VRF_SKEY} --tz UTC"
+  EPOCHS=$(sqlite3 "$BLOCKLOG_DB" "SELECT replace(group_concat(DISTINCT epoch ORDER BY epoch ASC), ',', ' ') AS EPOCHS FROM blocklog;")
```

Comment:
Invalid SQL, change to:
```
EPOCHS=$(sqlite3 "$BLOCKLOG_DB" "SELECT replace(group_concat(DISTINCT epoch), ',', ' ') AS EPOCHS FROM blocklog ORDER BY epoch ASC;")
```

Reply:
I tried, but epochs comes partially ordered ASC 

>464 465 466 467 468 469 470 471 472 473 474 475 476 230 252 274 282 299 311 312 316 317 318 319 320 321 322 323 324 325 326 327 328 329 330 331 332 333 334 335 339 380 387 391 392 393 394 395 396 397 398 399 400 401 402 403 404 406 407 408 409 410 411 412 413 414 415 416 417 418 419 420 421 422 423 424 425 426 427 428 429 430 431 432 433 434 435 436 437 438 439 440 441 442 443 444 445 446 447 448 449 450 451 452 453 454 455 456 457 458 459 460 461 462 463 477 478 479 480 481 482 483 484 485 486 487 488 489 490 491 492 493 494 495 496 497 498 499 500 501

do you agree with a solution like this?

>EPOCHS=$(sqlite3 "$BLOCKLOG_DB" "SELECT group_concat(epoch, ' ') AS EPOCHS FROM (SELECT DISTINCT epoch FROM blocklog ORDER BY CAST(epoch AS INTEGER) ASC);")

Reply:
Also, like you changed code to, remove replace and just add ' ' as separator to group_concat. But you should be able to order by epoch right away instead of casting to integer.

Reply:
Another option to distinct is to use GROUP BY, it would then be something similar to this.
```
EPOCHS=$(sqlite3 "$BLOCKLOG_DB" "SELECT group_concat(epoch, ' ') AS EPOCHS FROM blocklog GROUP BY epoch ORDER BY epoch ASC;")
```

Reply:
Ok, many comments but had to test it my self as sqlite isn't my goto database. I updated example in previous comment with what works best for me. 

Ie call group_concat on subquery as it will otherwise not work correctly with distinct.

Reply:
while "select epoch FROM blocklog GROUP BY epoch ORDER BY epoch ASC" returns distinct epochs

the combnation of group_concat func and GROUP BY makes return all the epochs from blocklog

I propose this:

EPOCHS=$(sqlite3 "$BLOCKLOG_DB" "SELECT group_concat(epoch, ' ') AS EPOCHS FROM (SELECT DISTINCT epoch FROM blocklog ORDER BY epoch ASC);")

Reply:
about  PRAGMA table_info(blocklog);
yes, epoch is integer

Reply:
You mention group by but example you posted use use distinct. Group by or distinct should both produce the same result so happy with whatever you like best.

Reply:
If you are referring to me using group by, I deleted that comment as it was just wrong. Please see my first example SQL as this is now updated.

Reply:
Hello @Scitz0 

thank you, I replaced with

>EPOCHS=$(sqlite3 "$BLOCKLOG_DB" "SELECT group_concat(epoch,' ') FROM (SELECT DISTINCT epoch FROM blocklog ORDER BY epoch);")

and just pushed the change